include/alpaka/block/sync/BlockSyncBarrierOacc.hpp

```diff
@@ -0,0 +1,48 @@
+/* Copyright 2020 Jeffrey Kelling
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#ifdef ALPAKA_ACC_ANY_BT_OACC_ENABLED
+
+#if _OPENACC < 201306
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
+#endif
+
+#include <alpaka/block/sync/Traits.hpp>
+
+namespace alpaka
+{
+    //#############################################################################
+    //! The OpenACC barrier block synchronization.
+    class BlockSyncBarrierOacc
+    {
+    public:
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST BlockSyncBarrierOacc() = default;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST BlockSyncBarrierOacc(BlockSyncBarrierOacc const &) = delete;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST BlockSyncBarrierOacc(BlockSyncBarrierOacc &&) = delete;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST auto operator=(BlockSyncBarrierOacc const &) -> BlockSyncBarrierOacc & = delete;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST auto operator=(BlockSyncBarrierOacc &&) -> BlockSyncBarrierOacc & = delete;
+        //-----------------------------------------------------------------------------
+        ~BlockSyncBarrierOacc() = default;
+
+        std::uint8_t mutable m_generation = 0u;
+        // NVHPC 20.7: initializer causes warning:
+        // NVC++-W-0155-External and Static variables are not supported in acc routine - _T139951818207704_37530
+        int mutable m_syncCounter[4] {0,0,0,0};
+        int mutable m_result;
```

Comment:
Is it a performance degradation if we initialize this value? Especially because all other members have default initializers.

Reply:
It is initialized before each use (by https://github.com/jkelling/alpaka/blob/openacc/include/alpaka/ctx/block/CtxBlockOacc.hpp#L214 ). After each us, it should be considered uninitalized again, so I do not see why it should be initialized even from a semantics standpoint.

Reply:
I know it's initialized later. But the question is: could this type be safer by having all members default initialized on construction without a performance penalty? If the `int mutable m_result = 0;` does not hurt, I would add it.

Reply:
Variables should not be initialized at all cost, because it effectively mutes use-before-init compiler warnings. If I do not actually want to rely on it having this value before first using it, I would consider initialization bad style.

Reply:
Btw. I do not know if there would be runtime cost, there might be on a GPU. This is more of a style /semantics issue for me.

Reply:
I would argue the store to this variable will be eliminated by a the compiler because it can see the store afterwards with the value you need in your algorithm later. This is not a style issue, but a safety issue. That's why it is also in the core guidelines: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Pro-type-memberinit

Reply:
Eliminating the store affects only the performance aspect.

Initializing it in the ctor is more likely to hide a bug: This variable is reused by subsequent invocations of that algorithm. If one forgot to init it at the beginning of the function, you would have a bug.

I usually agree with initializing things in ctors, that is what they are for, but this is the exception to the rule.

Reply:
I am with @jkelling here.